jetty-util/src/main/java/org/eclipse/jetty/util/IncludeExcludeSet.java

```diff
@@ -210,11 +211,72 @@ public void clear()
     @Override
     public String toString()
     {
-        return String.format("%s@%x{i=%s,ip=%s,e=%s,ep=%s}",this.getClass().getSimpleName(),hashCode(),_includes,_includePredicate,_excludes,_excludePredicate);
+        return String.format("%s@%x{i=%s,ip=%s,e=%s,ep=%s}",this.getClass().getSimpleName(),hashCode(),
+            _includes,
+            _includePredicate==_includes?"SELF":_includePredicate,
+            _excludes,
+            _excludePredicate==_excludes?"SELF":_excludePredicate);
     }
 
     public boolean isEmpty()
     {
         return _includes.isEmpty() && _excludes.isEmpty();
     }
+
+
+    /**
+     * Combine the results of two {@link IncludeExcludeSet}s with an "and" operation.  The items must be
+     * included by xSet AND ySet, and excluded from neither.   If a sets inclusions are empty, then all items
+     * are considered to be included.
+     * @param xSet The set that param x is tested against
+     * @param x An item to test against xSet
+     * @param ySet The set that param y is tested against
+     * @param y A supplier of an item to test against ySet, that is executed only of x is not excluded from xSet
+     * @param <XS> The type of xSet items
+     * @param <XP> The type of xSet predicates
+     * @param <YS> The type of ySet items
+     * @param <YP> The type of ySet predicates
+     * @return True only if the items are included and not excluded from their respected sets
+     */
+    public static <XS, XP, YS, YP> boolean and(IncludeExcludeSet<XS, XP> xSet, XP x, IncludeExcludeSet<YS,YP> ySet, Supplier<YP> y)
```

Comment:
Why can't the IncludeExcludeSet be a `java.util.function.Predicate` itself?
Then we can just use the `Predicate.and`, `Predicate.or` and `Predicate.negate` methods to accomplish this easier.

Reply:
why are you testing 2 different items in an `or`?

Reply:
``` java
return includeExcludeSetFoo.or(includeExcludeSetBar).test(item);`
```

Reply:
because the test is true IFF (x is included xSet OR y is included ySet) && NOT(x is excluded from xSet) && NOT(y is excluded from ySet)
The two sets are both predicates, but they are over different types.  Hence there are two items.

Reply:
Too difficult to understand. See above for simpler alternative.